model/flow/sealing_segment.go

```diff
@@ -42,45 +71,153 @@ func (segment *SealingSegment) Lowest() *Block {
+	for _, block := range segment.Blocks {
+		err := builder.AddBlock(block)
+		if err != nil {
+			return fmt.Errorf("invalid segment: %w", err)
+		}
+	}
+	_, err := builder.SealingSegment()
+	if err != nil {
+		return fmt.Errorf("invalid segment: %w", err)
+	}
+	return nil
+}
+
 var (
-	ErrSegmentMissingSeal        = fmt.Errorf("sealing segment failed sanity check: highest block in segment does not contain seal for lowest")
-	ErrSegmentBlocksWrongLen     = fmt.Errorf("sealing segment failed sanity check: must have atleast 2 blocks")
+	ErrSegmentMissingSeal        = fmt.Errorf("sealing segment failed sanity check: missing seal referenced by segment")
+	ErrSegmentBlocksWrongLen     = fmt.Errorf("sealing segment failed sanity check: non-root sealing segment must have at least 2 blocks")
 	ErrSegmentInvalidBlockHeight = fmt.Errorf("sealing segment failed sanity check: blocks must be in ascending order")
 	ErrSegmentResultLookup       = fmt.Errorf("failed to lookup execution result")
-	ErrInvalidRootSegmentView    = fmt.Errorf("invalid root sealing segment block view")
+	ErrSegmentSealLookup         = fmt.Errorf("failed to lookup seal")
+	ErrSegmentInvalidRootView    = fmt.Errorf("invalid root sealing segment block view")
 )
 
+// GetResultFunc is a getter function for results by ID.
+type GetResultFunc func(resultID Identifier) (*ExecutionResult, error)
+
+// GetSealByBlockIDFunc is a getter function for seals by block ID, returning
+// the latest seals incorporated as of the given block.
+type GetSealByBlockIDFunc func(blockID Identifier) (*Seal, error)
+
+// SealingSegmentBuilder is a utility for incrementally building a sealing segment.
 type SealingSegmentBuilder struct {
-	resultLookup    func(resultID Identifier) (*ExecutionResult, error)
+	// access to storage to read referenced by not included resources
+	resultLookup        GetResultFunc
+	sealByBLockIDLookup GetSealByBlockIDFunc
+	// keep track of resources included in payloads
 	includedResults map[Identifier]struct{}
-	blocks          []*Block
-	results         []*ExecutionResult
+	// resources to include in the sealing segment
+	blocks      []*Block
+	results     []*ExecutionResult
+	latestSeals map[Identifier]Identifier
+	firstSeal   *Seal
 }
 
-// AddBlock appends block to blocks
+// AddBlock appends a block to the sealing segment under construction.
 func (builder *SealingSegmentBuilder) AddBlock(block *Block) error {
-	//sanity check: block should be 1 height higher than current highest
+	// sanity check: block should be 1 height higher than current highest
 	if !builder.isValidHeight(block) {
 		return fmt.Errorf("invalid block height (%d): %w", block.Header.Height, ErrSegmentInvalidBlockHeight)
 	}
+	blockID := block.ID()
+
+	// construct a list of missing result IDs referenced by this block
+	missingResultIDs := make(map[Identifier]struct{})
 
-	// cache results in included results
+	// for the first (lowest) block, if it contains no seal, store the latest
```

Comment:
here we are assuming that we will add blocks in ascending order, maybe we shouldn't do this? 
We can wait for first seal calculation right before returning `SealingSegment`